teps/0061-allow-custom-task-to-be-embedded-in-pipeline.md

```diff
@@ -0,0 +1,239 @@
+---
+status: proposed
+title: Allow custom task to be embedded in pipeline
+creation-date: '2021-03-18'
+last-updated: '2021-03-27'
+authors:
+- '@Tomcli'
+- '@litong01'
+- '@ScrapCodes'
+---
+
+# TEP-0061: Allow custom task to be embedded in pipeline
+
+<!-- toc -->
+- [Summary](#summary)
+- [Motivation](#motivation)
+  - [Goals](#goals)
+  - [Non-Goals](#non-goals)
+  - [Use Cases (optional)](#use-cases-optional)
+- [Requirements](#requirements)
+- [Proposal](#proposal)
+  - [Notes/Caveats (optional)](#notescaveats-optional)
+  - [Risks and Mitigations](#risks-and-mitigations)
+  - [User Experience (optional)](#user-experience-optional)
+  - [Performance (optional)](#performance-optional)
+- [Design Details](#design-details)
+- [Test Plan](#test-plan)
+- [Design Evaluation](#design-evaluation)
+- [Drawbacks](#drawbacks)
+- [Alternatives](#alternatives)
+- [Infrastructure Needed (optional)](#infrastructure-needed-optional)
+- [Upgrade &amp; Migration Strategy (optional)](#upgrade--migration-strategy-optional)
+- [References (optional)](#references-optional)
+<!-- /toc -->
+
+## Summary
+
+Tektoncd/Pipeline currently allows custom task to be referenced in pipeline
+resource specification file using [`taskRef`](https://github.com/tektoncd/community/blob/main/teps/0002-custom-tasks.md).
+That feature allows a custom task to be submitted to kubernetes along
+the submission of the [Tektoncd/pipeline](https://github.com/tektoncd/pipeline),
+however, the submission of a custom task is a separate request to Kubernetes.
+If multiple custom tasks use same name, to both Kubernetes and Tektoncd/Pipeline,
+they will be treated as the same task, this behavior can have unintended
+consequences when Tektoncd/Pipeline gets used as a backend with multiple users.
+This problem becomes even greater when new users follow documents such as
+`Get started` which often use same name for task and pipeline. In this environment
+multiple users will step on each other's toes, and produced unintended results.
+
+## Motivation
+
+It is natural for a user to follow ways such as a [Kubernetes Deployment](https://kubernetes.io/docs/concepts/workloads/controllers/deployment/), ReplicaSet, StatefulSet and also
```

Comment:
This jumps directly into a general assertion that there are natural or unnatural ways for users to work with Kubernetes. This is useful context but I think it would be really helpful if we could nail down a specific motivation too.

From your experiences using custom tasks can you include here a description of the exact problem you and your users face?  Ideally not generalized over an entire problem area but very concretely: `User X does A, User Z does B, these actions collide because C, we are then motivated to add Custom Task Specs to fix it by ...`

Reply:
We will move our Kubeflow pipeline use case over here since that's the exact problem we are facing